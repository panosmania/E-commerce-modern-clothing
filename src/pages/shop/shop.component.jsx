import React, { useEffect, lazy, Suspense } from "react";
import { Route } from "react-router-dom";
import { createStructuredSelector } from "reselect";
import { connect } from "react-redux";

//import CollectionsOverview from "../../components/collections-overview/collections-overview.component";
//import CollectionPage from "../collection/collection.component";

import { fetchCollectionsStart } from "../../redux/shop/shop.actions";
import {
  selectCollectionFetching,
  selectIsCollectionsLoaded,
} from "../../redux/shop/shop.selectors";

import WithSpinner from "../../components/with-spinner/with-spinner.component";

import Spinner from "../../components/spinner/spinner.component";

const CollectionsOverview = lazy(() =>
  import("../../components/collections-overview/collections-overview.component")
);

const CollectionPage = lazy(() => import("../collection/collection.component"));

const CollectionsOverviewWithSpinner = WithSpinner(CollectionsOverview);
const CollectionPageWithSpinner = WithSpinner(CollectionPage);

export const ShopPage = ({
  fetchCollectionsStart,
  match,
  isCollectionFetching,
  isCollectionsLoaded,
}) => {
  useEffect(() => {
    fetchCollectionsStart();
  }, [fetchCollectionsStart]);

  return (
    <div className="shop-page">
      <Suspense fallback={<Spinner />}>
        <Route
          exact
          path={`${match.path}`}
          render={(props) => (
            <CollectionsOverviewWithSpinner
              isLoading={isCollectionFetching}
              {...props}
            />
          )}
        />
        <Route
          path={`${match.path}/:collectionId`}
          render={(props) => (
            <CollectionPageWithSpinner
              isLoading={!isCollectionsLoaded}
              {...props}
            />
          )}
        />
      </Suspense>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  isCollectionFetching: selectCollectionFetching,
  isCollectionsLoaded: selectIsCollectionsLoaded,
});

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionsStart: () => dispatch(fetchCollectionsStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ShopPage);
