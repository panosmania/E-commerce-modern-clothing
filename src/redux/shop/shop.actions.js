import ShopActionTypes from "./shop.types";

import {
  firestore,
  convertCollectionsSnapshotToMap
} from "../../firebase/firebase.utils";

export const fetchCollectionsStart = () => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_START
});

export const fetchCollectionsSuccess = collectionsMap => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_SUCCESS,
  payload: collectionsMap
});

export const fetchCollectionsFailure = errorMessage => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_FAILURE,
  payload: errorMessage
});

export const fetchCollectionsStartAsync = () => {
  return dispatch => {
    const collectionRef = firestore.collection("collections");
    dispatch(fetchCollectionsStart());

    //get data
    collectionRef
      .get()
      .then(async snapshot => {
        //console.log(snapshot);
        const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
        //console.log(collectionsMap);

        dispatch(fetchCollectionsSuccess(collectionsMap));
      })
      .catch(error => dispatch(fetchCollectionsFailure(error.message)));
  };
};
