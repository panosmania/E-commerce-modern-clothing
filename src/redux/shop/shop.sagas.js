import { takeLatest, call, put, all } from "redux-saga/effects";
import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";

import {
  fetchCollectionsSuccess,
  fetchCollectionsFailure,
} from "./shop.actions";

import ShopActionTypes from "./shop.types";

export function* fetchCollectionsStartAsync() {
  //yield console.log("Iam fired");

  try {
    const collectionRef = firestore.collection("collections");
    const snapshot = yield collectionRef.get();
    const collectionsMap = yield call(
      convertCollectionsSnapshotToMap,
      snapshot
    );
    yield put(fetchCollectionsSuccess(collectionsMap));
  } catch (error) {
    yield put(fetchCollectionsFailure(error.message));
  }

  //get data
  //   collectionRef
  //     .get()
  //     .then(async (snapshot) => {
  //       //console.log(snapshot);
  //       const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
  //       //console.log(collectionsMap);

  //       dispatch(fetchCollectionsSuccess(collectionsMap));
  //     })
  //     .catch((error) => dispatch(fetchCollectionsFailure(error.message)));
}

export function* fetchCollectionsStart() {
  yield takeLatest(
    ShopActionTypes.FETCH_COLLECTIONS_START,
    fetchCollectionsStartAsync
  );
}

export function* shopSagas() {
  yield all([call(fetchCollectionsStart)]);
}
