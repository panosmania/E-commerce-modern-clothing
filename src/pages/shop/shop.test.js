import { shallow } from "enzyme";
import React from "react";
import toJson from "enzyme-to-json";

import { ShopPage } from "./shop.component";

describe("ShopPage", () => {
  let wrapper;
  let mockFetchCollectionsStart;
  let mockMatch;
  const isCollectionFetching = true;
  const isCollectionsLoaded = true;

  beforeEach(() => {
    mockMatch = {
      path: "",
    };
    mockFetchCollectionsStart = jest.fn();

    const mockProps = {
      isCollectionFetching: isCollectionFetching,
      isCollectionsLoaded: isCollectionsLoaded,
      match: mockMatch,
      CollectionsStart: mockFetchCollectionsStart,
    };

    wrapper = shallow(<ShopPage {...mockProps} />);
  });

  it("should render ShopPage component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render ShopPage component at least >0", () => {
    expect(mockFetchCollectionsStart).not.toHaveBeenCalled();
  });
});
