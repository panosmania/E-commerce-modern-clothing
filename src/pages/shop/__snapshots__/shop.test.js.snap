// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage should render ShopPage component 1`] = `
<div
  className="shop-page"
>
  <Suspense
    fallback={<Spinner />}
  >
    <Route
      exact={true}
      path=""
      render={[Function]}
    />
    <Route
      path="/:collectionId"
      render={[Function]}
    />
  </Suspense>
</div>
`;
